py4j-python/src/py4j/tests/java_dir_test.py

```diff
@@ -0,0 +1,162 @@
+                      "method8",
+                      "method9",
+                      "method10", # overloaded
+                      "method11",
+                      "getList",
+                      "getField1",
+                      "setField1",
+                      "getStringArray",
+                      "getIntArray",
+                      "callHello",
+                      "callHello2",
+                      "static_method",
+
+                      # From Object
+                      "getClass",
+                      "hashCode",
+                      "equals",
+                      "toString",
+                      "notify",
+                      "notifyAll",
+                      "wait"
+                      ])
+
+ExampleClassStatics = sorted([
+                              "StaticClass",
+                              "static_field",
+                              "static_method"
+                              ])
+
+@contextmanager
+def example_app_process():
+    p = start_example_app_process()
+    try:
+        yield p
+    finally:
+        p.join()
+
+@contextmanager
+def gateway(*args, **kwargs):
+    g = JavaGateway(
+        gateway_parameters=GatewayParameters(*args, auto_convert=True, **kwargs))
+    lineSep = g.jvm.System.lineSeparator()
+    try:
+        yield g
+        # Call a dummy method to make sure we haven't corrupted the streams
+        eq_(lineSep, g.jvm.System.lineSeparator())
+    finally:
+        g.shutdown()
+
+def test_dir_object():
+    with example_app_process():
+        with gateway() as g:
+            ex = g.getNewExample()
+            eq_(sorted(dir(ex)), ExampleClassMethods)
+
+def test_dir_object_fields():
+    with example_app_process():
+        with gateway(auto_field=True) as g:
+            ex = g.getNewExample()
+            eq_(sorted(dir(ex)), sorted(ExampleClassMethods + ExampleClassFields))
+
+def test_dir_object_shows_manually_called_after_dir():
+    with example_app_process():
+        with gateway() as g:
+            ex = g.getNewExample()
+            eq_(sorted(dir(ex)), ExampleClassMethods)
+            try:
+                ex.does_not_exist_in_example()
+                raise AssertionError("Method should not have succeeded")
+            except:
+                pass
+            # Make sure the manually called method now shows up
+            eq_(sorted(dir(ex)), sorted(ExampleClassMethods + ['does_not_exist_in_example']))
+
+def test_dir_object_shows_manually_called_before_dir():
+    with example_app_process():
+        with gateway() as g:
+            ex = g.getNewExample()
+            try:
+                ex.does_not_exist_in_example()
+                raise AssertionError("Method should not have succeeded")
+            except:
+                pass
+            # Make sure the manually called method now shows up
+            eq_(sorted(dir(ex)), sorted(ExampleClassMethods + ['does_not_exist_in_example']))
```

Comment:
Good question. I am not sure of the right answer, I don't think it is great for usability, but the __dir__ in this case is reflecting the internal state of the JavaObject. The JavaObject is storing in self._methods the JavaMember regardless of whether the call succeeds. What could potentially be done is that once self._populate_fields() has been called error early on calling methods not already in self._methods. However I resisted erroring early in the first instance as I was concerned that I didn't affect existing behaviour of Py4J. But once that member is added to self._methods I didn't want to then ignore it. 

Perhaps the test would be more accurate as something like:

```python
            assert (sorted(dir(ex)) == sorted(ExampleClassMethods + ['does_not_exist_in_example'])) or (sorted(dir(ex)) == sorted(ExampleClassMethods)
```

As that would more accurately represent the contract I would like to see maintained going forward.